libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,332 @@
+    step (SearchArgs Z _ bLteA {bNonZero}) _ = absurd . succNotLTEzero $ lteTransitive bNonZero bLteA
+    step (SearchArgs _ Z _ {bNonZero}) _ = absurd $ succNotLTEzero bNonZero
+    step (SearchArgs (S a) (S b) bLteA {bNonZero}) rec with (divMod (S a) (S b))
+        | Fraction (S a) (S b) q FZ prf =
+            let sbIsFactor = the (plus q (mult b q) = S a) $
+                    rewrite multCommutative b q in
+                    rewrite sym $ multRightSuccPlus q b in
+                    replace {P = \x => x = S a} (plusZeroRightNeutral (q * S b)) prf
+                skDividesA = CofactorExists q (lteTransitive bNonZero bLteA) sbIsFactor
+                skDividesB = selfFactor (S b)
+                greatest = \q', (CommonFactorExists q' _ qfb) => qfb
+            in
+            (S b ** MkGCD (CommonFactorExists (S b) skDividesA skDividesB) greatest)
+
+        | Fraction (S a) (S b) q (FS r) prf =
+                let rLtSb = lteSuccRight $ elemSmallerThanBound r
+                    qGt1 = the (LTE 1 q) $ case q of
+                        Z => absurd . notLteAndGt (S $ finToNat r) b (elemSmallerThanBound r) $
+                            replace {P = LTE (S b)} (sym prf) bLteA
+                        (S k) => LTESucc LTEZero
+                    smaller = the (LTE (S (S (plus b (S (finToNat r))))) (S (plus a (S b)))) $
+                        rewrite plusCommutative a (S b) in
+                        LTESucc . LTESucc . addLteLeft . fromLteSucc $ lteTransitive (elemSmallerThanBound $ FS r) bLteA
+                    (f ** MkGCD (CommonFactorExists f prfSb prfRem) greatestSbSr) =
+                        rec (SearchArgs (S b) (S $ finToNat r) rLtSb) smaller
+                    prfSa = the (Factor (S a) f) $
+                        rewrite sym prf in
+                        rewrite multCommutative q (S b) in
+                        plusFactor (multNAlsoFactor prfSb q) prfRem
+                    greatest = \q', (CommonFactorExists q' qfa qfb) =>
+                        let sbfqSb =
+                                the (Factor (q * S b) (S b)) $
+                                rewrite multCommutative q (S b) in
+                                multFactor (S b) q
+                            rightPrf = minusFactor {a = q * S b} {b = S (finToNat r)}
+                                (rewrite prf in qfa)
+                                (factorTransitive sbfqSb qfb)
+                        in
+                        greatestSbSr q' (CommonFactorExists q' qfb rightPrf)
+                in
+                (f ** MkGCD (CommonFactorExists f prfSa prfSb) greatest)
+
+||| An implementation of Euclidean Algorithm for computing greatest common
+||| divisors. It is proven correct and total; returns a proof that computed
+||| number actually IS the GCD. Unfortunately it's very slow, so improvements
+||| in terms of efficiency would be welcome.
+gcd : (a, b : Nat) -> {auto aok : LTE 1 a} -> {auto bok : LTE 1 b} -> (f : Nat ** GCD a b f)
```

Comment:
There is a `gcd` defined in [Prelude.Nat](https://github.com/idris-lang/Idris-dev/blob/0417c53fb6b91fc0692245f269b98a3bd4820136/libs/prelude/Prelude/Nat.idr#L419)

Reply:
True. But that definition does not provide any proof that the returned number is indeed what it claims to be. That is the whole point of this module: to define what a factor is and start proving things about factors and factorisation.

Reply:
Yes, my point is just that there is a name collision.

Reply:
I thought about naming it `gcf` for greatest common factor. That would be along with the naming convention in this module, but on the other hand it would *not* be along with what people expect that function to be named. Type of this function is different, so in some cases (when you supply implicit arguments explicitly) calls will be unambiguous. Otherwise you'll have to qualify it. I don't think it's too bad…

Reply:
The other `gcd` function only stipulates that the inputs can't both be zero. Could that work here? I've looked at a few expositions of the algorithm, and nobody ever spends much time discussing the nature of the inputs. It would certainly be cleaner if there weren't `LTE 1 n` proofs floating around all over, but I don't know how necessary they are.

Reply:
Zeros are problematic, because according to this definition at least, everything is a factor of zero. This destroys some nice properties, like that a factor is always less than or equal to the number it divides. Also there is consideration that according to this setup, zero is a factor of zero, which implicitly assumes that you can actually divide by zero. Also when you want to implement rational numbers, that non-zero property comes in handy, because all ratios are either positive or negative and you just add a special constructor for zero. This way you avoid the problem `ZZ` has, that `0` and `-0` are distinct values, which messes up the `(=)` proofs. Also the requirement that divisor must be non-zero is graciously provided by this arrangement.

On the other hand, as I review the code, I see that the non-zero property is not used that often (mostly is just passed forward as is or with some alteration). So probably we could require non-zero-ness for those proofs only, and we could require it explicitly fro rational numbers as well. I'm currently looking into the possibilities of removing that constraint and it seems possible.